x-pack/plugins/upgrade_assistant/server/lib/reindexing/worker.ts

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { CallCluster } from 'src/legacy/core_plugins/elasticsearch';
+import { Server } from 'src/server/kbn_server';
+import { SavedObjectsClient } from 'src/server/saved_objects';
+
+import { ReindexSavedObject } from '../../../common/types';
+import { ReindexService, reindexServiceFactory } from './reindex_service';
+
+const POLL_INTERVAL = 30000;
+
+const LOG_TAGS = ['upgrade_assistant', 'reindex_worker'];
+
+/**
+ * A singleton worker that will coordinate two polling loops:
+ *   (1) A longer loop that polls for reindex operations that are in progress. If any are found, loop (2) is started.
+ *   (2) A tighter loop that pushes each in progress reindex operation through ReindexService.processNextStep. If all
+ *       updated reindex operations are complete, this loop will terminate.
+ *
+ * The worker can also be forced to start loop (2) by calling forceRefresh(). This is done when we know a new reindex
+ * operation has been started.
+ *
+ * This worker can be ran on multiple nodes without conflicts or dropped jobs. Reindex operations are locked by the
+ * ReindexService and if any operation is locked longer than the ReindexService's timeout, it is assumed to have been
+ * locked by a node that is no longer running (crashed or shutdown). In this case, another node may safely acquire
+ * the lock for this reindex operation.
+ */
+export class ReindexWorker {
+  private static workerSingleton?: ReindexWorker;
+  private continuePolling: boolean = false;
+  private updateOperationLoopRunning: boolean = false;
+  private inProgressOps: ReindexSavedObject[] = [];
+  private readonly reindexService: ReindexService;
+
+  private processNextStep: (reindexOp: ReindexSavedObject) => Promise<ReindexSavedObject>;
+
+  constructor(
+    client: SavedObjectsClient,
+    callWithInternalUser: CallCluster,
+    private readonly log: Server['log']
+  ) {
+    if (ReindexWorker.workerSingleton) {
+      throw new Error(`More than one ReindexWorker cannot be created.`);
+    }
+
+    this.reindexService = reindexServiceFactory(client, callWithInternalUser);
```

Comment:
We should still be using `callWithRequest` here. In most deployments, the default `kibana` user won't have the required privileges to interact with indices other than `.kibana*`

Reply:
Actually, I think I see the dilemma here since we are processing these on the server in a queue.

Reply:
We talked over this with @kobelb and decided that we will cache the user's credentials in memory for the worker to use.

If the worker does not have the credentials it needs, we will not continue to process the Reindex Operation (putting it into a "paused" state) until these credentials are re-populated by an authenticated user visiting the Upgrade Assistant and clicking a "resume" button.